### PR TITLE
feat(machine): Phase 4 — cross-tab consistency, narrow-viewport pass, dead-code removal

### PR DESCRIPTION
**Pull request by 2witstudios**
Final phase of the Machine page rebuild (Terminal — GA Release epic). Phase 3 landed all four tabs (Terminal / Code / Diff / Settings), but they were built by four agents in isolation, so this is the pass that makes them one surface. Task list: `zycgqurfcorlh7urb292b4fa`.

No tab's core behavior and no backend route changed. Admin + `CODE_EXECUTION_ENABLED` gating untouched.

## 1. Cross-tab consistency + loading/error/empty states

Four agents produced four different treatments of the same three moments: a bordered spinner in one pane, a bare `Loader2` in another, three different empty-row layouts, error text that sometimes had a retry and sometimes didn't.

**`tabs/tab-states.tsx`** is those moments factored once. The split is by **place**, not by tab — a state in an inner sidebar is a compact row that sits among tree rows (`SidebarLoading`, `SidebarNotice`), and a state in a main pane is a centred block that owns the area (`PaneLoading`, `PaneNotice`) — so every tab reaches for both. One `Spinner`. One `tone` (`muted` for a state of the world, `destructive` for an actual failure).

What is deliberately **not** unified is the words. The Terminal tab's connect state, the Code tab's "this branch hasn't been checked out yet" and the Diff tab's "No uncommitted changes on main" are three different facts; callers still pass their own copy. Two of them got sharper:

- **Cold boot vs. fast reattach.** A reattach answers in well under a second; a cold Sprite has to boot, and a bare "Connecting…" sitting for ten seconds reads as a hang. The socket doesn't say which is happening, but the clock does — past 1.5s the message becomes "Starting the sandbox — a cold boot takes a few seconds…".
- Errors a refetch could fix now carry a **Retry**: the diff list (wired to the same revalidation the header refresh runs), the file listing, and the settings load.

**`tabs/TabSidebar.tsx`** is the other half: one shell behind the Terminal tab's session tree, the Code tab's file explorer and the Diff tab's branch tree. Same width, same `border-border` chrome, same header treatment, same row hover/selected states — they can't drift apart again because there is now only one of them.

## 2. Narrow viewports

Same convention the app already uses in `Layout.tsx`, through the same `useMobile()` hook.

- **Inner sidebar → Sheet.** A 16rem column beside a pane leaves a phone with a sliver, and a sliver of Monaco or xterm is worse than nothing. Below the breakpoint the sidebar goes behind a header button and the pane takes the full width. The sidebar body is a render prop taking `close`, because only the caller knows which of its clicks *navigates* (picking a session/branch/file — dismiss, the thing you opened is behind the sheet) and which merely *expands* (keep it open, the user is still drilling down).
- **Pane splitting.** Two columns at 375px give each terminal ~180px. On narrow viewports only the **active pane is visible**; the split controls are hidden and the store's layout is untouched, so a desktop split comes back laid out on the next wide render. The inactive panes stay **mounted and hidden**, never unmounted — see the review round below. A pane strip appears when, and only when, the collapse actually hid something, so those panes are never silently unreachable. Pane controls also stay visible on touch instead of waiting for a hover that never comes.
- **Tab triggers** go icon-only below `sm` (`aria-label` carries the name the hidden label no longer contributes). The diff **scope toggle** stacks under the path and scrolls sideways rather than compressing three labels to ellipsis, and Monaco **diffs render inline** instead of side-by-side.

## 3. Dead code

- `ComingSoonTab.tsx` — all three placeholders were replaced in Phase 3; `git grep` confirms nothing imported it.
- `workspace/EmptyState.tsx` — absorbed into `tab-states`; its copy still pointed users at the "navigator", a surface that no longer exists.
- `terminal/types.ts` — `HistoryEntry`/`TerminalSession`, unimported remnants of the pre-rebuild command-history terminal.

## Review round (bot reviews + self-review)

- **Hidden mobile panes must stay mounted** (Codex). The first cut of the narrow-viewport collapse rendered the active pane alone, unmounting the others — and unmounting an `XtermTerminal` emits `agent-terminal:disconnect`, which nulls the session's `closedFn` and arms the idle reap. An agent finishing off-screen would lose its final output and exit code, and returning to that pane would cold-start a fresh PTY. Inactive panes are now hidden, not unmounted.
- **…and hidden by `visibility`, not `display`** (self-review of that fix). A mobile pane can *mount* while inactive, and xterm measures its character cell from the DOM at `open()`; in a `display:none` box that measurement is 0, `FitAddon` proposes no dimensions, and even the refit on re-show is a no-op — the pane would stay blank permanently. `visibility: hidden` keeps the box at its real size and keeps `offsetParent`/`clientWidth` truthy, which is what `XtermTerminal`'s own visibility gate checks.
- **`min-h-0` on the sidebar `ScrollArea`** (CodeRabbit). A flex child defaults to `min-height: auto`, so a long tree grew the sidebar past the pane instead of scrolling inside it. All three tabs had this before the consolidation; fixing it in the shared shell fixes all three.
- Self-review also found: an **empty control chip** rendering on a lone phone pane (the chip is `opacity-100` on touch — with nothing to split into and nothing to close, an empty bordered box sat in the terminal's corner permanently); the focus accent reading a different pane id than the one actually shown in the fallback case; a `useCallback` in the Terminal tab that a per-render render-prop object was quietly invalidating; and the diff's refresh button being squeezed by the scrolling scope toggle beside it. All fixed.

## Verification

`bun run typecheck` (apps/web) and eslint clean; CI's Lint & TypeScript check is green. 120 tests pass in the terminal suite, including new colocated tests for `TabSidebar` (desktop vs. sheet; expanding keeps the sheet open, navigating closes it) and `TerminalPanes` (active-pane-only collapse, hidden panes staying mounted, hidden-by-visibility-not-display, the pane strip as the way back). Existing tests were updated only where the shared copy changed.

The full local `apps/web` suite also shows 16 failures across 4 files (`activity-tools`, `admin-role-version`, `grouping`, `version-sdk-contract`). Those are DB-dependent tests that need `bun run test`'s database setup rather than a bare `vitest` run — none of them reference anything in this diff (which touches only `page-views/terminal/**`), and master's own CI Test Suite is green on `f4980cc55`.

Not verified by driving the real Machine page — it needs `CODE_EXECUTION_ENABLED` plus live Sprites. In particular the xterm-in-a-hidden-container reasoning above is derived from the code, not observed in a browser.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01482wpEk73wDRERkLbHrbnP

### CODE REVIEW

File: apps/web/src/components/layout/middle-content/page-views/terminal/workspace/TerminalPanes.tsx [hunk 1]
```diff
@@ -28,13 +30,49 @@ export default function TerminalPanes({ machineId, socket }: TerminalPanesProps)
...
+  if (isMobile) {
+    const active = panes.find((pane) => pane.id === activePaneId) ?? panes[0];
+    if (!active) return null;
+    return (
+      <div className="flex h-full flex-col bg-background">
+        {panes.length > 1 && (
+          <PaneStrip panes={panes} activePaneId={activePaneId} onSelect={(id) => selectPane(machineId, id)} />
+        )}
+        <div className="min-h-0 flex-1">
+          <TerminalPane {...paneProps(active)} canSplit={false} />
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/terminal/workspace/TerminalPanes.tsx]: Confirmed and fixed in `e14dacd2f`.

You're right about the mechanism. I traced it: unmounting `XtermTerminal` emits `agent-terminal:disconnect`, and `disconnectConnection` (apps/realtime/src/terminal/agent-terminal-handler.ts:434) sets `session.closedFn = () => {}` and arms the `DETACHED_IDLE_MS` reap. So a hidden pane's agent finishing off-screen drops its `closed` event, and once reaped the next connect takes the cold path and spawns a fresh PTY rather than replaying the completed run.

Inactive panes are now **hidden, not unmounted** — each pane renders into an `absolute inset-0` wrapper that gets `hidden` when it isn't the active one, so the socket connection and scrollback stay live. This is a state `XtermTerminal` already supports explicitly: it skips fits while the container is 0×0 (avoiding a bogus 0-wide resize to the PTY) and its ResizeObserver fires exactly once with correct dimensions when the pane is shown again — the same path a keep-alive terminal takes when its page stops being the active tab.

Covered by a new test in `TerminalPanes.test.tsx` ("the panes it hides stay MOUNTED") asserting both terminals remain mounted with one hidden.

File: apps/web/src/components/layout/middle-content/page-views/terminal/tabs/TabSidebar.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+              <ScrollArea className="min-h-0 flex-1">{children({ close: () => setSheetOpen(false) })}</ScrollArea>
+            </SheetContent>
+          </Sheet>
+        </div>
+      ) : (
+        <aside className="flex w-64 shrink-0 flex-col border-r border-border bg-background">
+          <div className="flex shrink-0 items-center justify-between border-b border-border px-3 py-2">
+            <span className={HEADER_LABEL}>{title}</span>
+          </div>
+          <ScrollArea className="flex-1">{children(DESKTOP_API)}</ScrollArea>
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/terminal/tabs/TabSidebar.tsx]: Fixed in `e14dacd2f` — the desktop `ScrollArea` is now `min-h-0 flex-1`.

Correct diagnosis: the `aside` is a column flex container, and a flex child defaults to `min-height: auto`, so the ScrollArea could grow to its content height and a long Machine tree would push past the pane instead of scrolling inside the sidebar. Worth noting this was inherited — all three tabs' sidebars had the bare `flex-1` before this PR consolidated them — so fixing it in `TabSidebar` fixes it for the Terminal, Code and Diff trees at once, which is exactly what the shared shell is for.

File: apps/web/src/components/layout/middle-content/page-views/terminal/workspace/TerminalPanes.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/terminal/workspace/TerminalPanes.tsx]: Correction to my previous reply, and a follow-up in `213642bba`.

I said a `display:none` terminal was an already-supported state. That is true only for the case the keep-alive path exercises — hiding a terminal that was **already opened at a real size**. It is not true here: on this surface a pane can *mount* while inactive, and xterm measures its character cell from the DOM inside `open()`. In a `display:none` box that measurement is 0, `FitAddon.proposeDimensions()` then returns undefined, and the ResizeObserver refit on re-show is a no-op — the pane would stay blank permanently.

Inactive panes are now hidden with `invisible` (`visibility: hidden`) instead. All panes are stacked at `absolute inset-0`, so a hidden one keeps the container's real box: the char measurement and the fit are correct, and `offsetParent`/`clientWidth` stay truthy — which is precisely what `XtermTerminal`'s own visibility gate tests before it fits. The PTY-preservation property your comment asked for is unchanged; only the hiding mechanism is.

New test ("a hidden pane is hidden by VISIBILITY, never by display") pins this so it can't silently regress back to `hidden`.
